pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,756 @@
+		klog.InfoS("Skip processing ExternalNode update as no changes for Interface[0]", "ExternalNode", klog.KObj(curEn))
+		return nil
+	}
+	preEEName, err := externalnode.GenExternalEntityName(preEn)
+	if err != nil {
+		return err
+	}
+	preIfName, preIPs, err := getHostInterfaceName(preEn.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	curEEName, err := externalnode.GenExternalEntityName(curEn)
+	if err != nil {
+		return err
+	}
+	curIfName, curIPs, err := getHostInterfaceName(curEn.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	if preIfName != curIfName {
+		klog.InfoS("Found interface name is changed", "preName", preIfName, "curName", curIfName)
+		if err = c.addInterface(curIfName, curEn.Namespace, curEEName, curIPs); err != nil {
+			return err
+		}
+		if err = c.deleteInterface(preIfName); err != nil {
+			return err
+		}
+	} else if (!reflect.DeepEqual(preIPs, curIPs)) || (preEEName != curEEName) {
+		klog.InfoS("Found interface configuration is changed", "preIPs", preIPs, "preExternalEntity", preEEName,
+			"curIPs", curIPs, "curExternalEntity", curEEName)
+		if err = c.addInterface(curIfName, curEn.Namespace, curEEName, curIPs); err != nil {
+			return err
+		}
+	}
+	c.syncedExternalNodes.Add(curEn)
+	// Notify the ExternalEntity event to NetworkPolicyController.
+	c.nodeUpdateNotifier.Notify(v1beta2.ExternalEntityReference{
+		Name:      curEEName,
+		Namespace: curEn.Namespace,
+	})
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalNode(key string) error {
+	klog.InfoS("Deleting ExternalNode", "key", key)
+	obj, exists, _ := c.syncedExternalNodes.GetByKey(key)
+	if !exists {
+		klog.InfoS("Skipping ExternalNode deletion as it hasn't been synced", "ExternalEntityKey", key)
+		return nil
+	}
+	en := obj.(*v1alpha1.ExternalNode)
+	ifName, _, err := getHostInterfaceName(en.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	if err = c.deleteInterface(ifName); err != nil {
+		return err
+	}
+	c.syncedExternalNodes.Delete(en)
+	return nil
+}
+
+func (c *ExternalNodeController) deleteInterface(ifName string) error {
+	hostIface, ifaceExists := c.ifaceStore.GetInterfaceByName(ifName)
+	if !ifaceExists {
+		klog.InfoS("Skipping deleting host interface since it doesn't exist ", "ifName", ifName)
+		return nil
+	}
+	klog.InfoS("Deleting interface", "ifName", ifName)
+	if err := c.removeOVSPortsAndFlows(hostIface); err != nil {
+		return err
+	}
+	c.ifaceStore.DeleteInterface(hostIface)
+	return nil
+}
+
+func (c *ExternalNodeController) createOVSPortsAndFlows(uplinkName, hostIFName, eeNamespace, eeName string, ips []string) (*interfacestore.InterfaceConfig, error) {
+	adapterConfig, err := util.GetUplinkConfig(hostIFName)
+	if err != nil {
+		return nil, err
+	}
+	if err = util.RenameInterface(hostIFName, uplinkName); err != nil {
+		return nil, err
+	}
+	success := false
+	defer func() {
+		if !success {
+			if err = util.RenameInterface(uplinkName, hostIFName); err != nil {
```

Comment:
I understand your point and Wenying and I have had a discussion on it. However, we did not find a way to avoid this loop of failure if we did get error in the defer clause. Retry will always start from the beginning rather than continue from where it failed last time. So you are right that we can get problem when we retry RenameInterface from hostIFName to uplinkName if the rollback fails. In my point, we may need to manually recover the setup if it did occur.

Reply:
At least we should add a comment about this, and even better to include it in the log message, e.g. sth. like "Manual cleanup is required".

Reply:
BTW, I remember NodeRouteController save the processing state to a cache, so retry knows where to start. But I know it can be complex and am not saying you must do that for ExternalNode.

Reply:
Thanks for the comment. I have added this to log message as suggested.